### ISSUE TITLE
Property "Craft\Shortlist_ItemModel.section" is not defined.

### ISSUE THREAD

**Issue by Moorst** [michaelrog/Shortlist#24]
Getting this error when trying to add item to a list. Not doing anything special, just trying to add an entry to the default list;

```
{% set item = craft.shortlist.item(entry.id) %}
{% if item.inList %}
<a class="link-shortlist" href="{{ item.removeActionUrl }}">Remove from shortlist</a>
{% else %}
<a class="link-shortlist" href="{{ item.addActionUrl }}">Add to shortlist</a>
{% endif %}
```

Any idea why?

**Comment by Moorst** [comment]
Realised this is an issue with the CacheFlag plugin rather than yourselves, opened up a ticket there instead. https://github.com/mmikkel/CacheFlag-Craft/issues/20.

**Comment by mmikkel** [comment]
The issue here is that in a `onElementSave` handler, the `ShortList_ItemModel` instance will return `"Entry"` for `$event->params['element']->elementType` (magic getter). Using the actual getter method – i.e. `$event->params['element']->getElementType()` – returns `"ShortList_Item"` as expected. Not really sure if this is an issue with Shortlist or Craft, tbh.

TL;DR – the issue is fixed in the [latest CacheFlag version](https://github.com/mmikkel/CacheFlag-Craft/releases/tag/v.1.1.11), but would be interesting to hear your input on the underlying issue @michaelrog ;)

**Comment by johndwells** [comment]
I believe we've tracked down the source of this issue.

`Shortlist_Item`, being an instance of `BaseModel` (with a corresponding `Shortlist_ItemRecord` for a db table), defines some attributes, including `elementId` and `elementType`, and these attributes (and database columns) are used to track what _element_ is saved as a short list item.

Problem is, `Shortlist_Item` is also an element itself - it's an instance of `BaseElementType` (which itself extends `BaseModel`). Therefore it has a protected property `Shortlist_Item::$elementType` which _should_ be returned by the  getter `BaseElementType::getElementType()`.

But since `elementType` has been defined as an attribute on the model, then it' s `BaseModel::getAttribute()` that wins the race in the magic getter war.  `BaseElementType::getElementType()` never has a chance to run.

I still haven't quite figured out why `Shortlist_Item::$elementType` is evaluating to `Entry`, but I would imagine it's somewhere deep inside the "default" behaviour of being _elemental_. At this point I'm not going to dig any further.

The fix therefore is "simple" in that it cannot have an attribute named `elementType`. For example I refactored the code to just `elType` and everything seemed to work.

For the record, we discovered this bug not through using CacheFlag, but Plus (https://github.com/fvaldes33/plus). So this is a big deal, it's a fundamental incompatability with Craft/element architecture. I realise that this version of Shortlist may potentially be "dead" in so much as @michaelrog is probably focused more on the C3-compatible version (right???), but this is a worthy bug to note, as it may still be relevant on that codebase.

**Comment by joelbradbury** [comment]
Only just noticed this thread but I can confirm John's diagnosis of the
root cause.

Originally the shortlist item was simply a model and only became it's own
element type once the field type support was added.

At the time there were some other similar issues around the nesting nature
of elements within elements and getting everything properly behaving with
the black box that is/was the elemental support of craft 2 while also being
usable within the twig templating without a bunch of extra hoops to jump
through.

The elType patch John mentioned should bypass most of the issues.

**Comment by johndwells** [comment]
Thanks @joelbradbury for chiming in! Yeah, swapping to `elType` seems to be working OK for us so far. 👍
